presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -500,12 +501,18 @@ public void partitionPage(Page page)
                 if (shouldReplicate) {
                     for (int i = 0; i < partitionBuffers.length; i++) {
                         partitionBuffers[i].addPosition(position);
+                        if (position == 0) {
```

Comment:
There are several problems here:
1) The page would be distributed into the partition buffers therefore each partition would only get a portion of the rows. But you're counting the whole page.
2) It's bad to put this counting into the loop body, as it would be called on every position. It is generally not desired to put unnecessary branching statements like `if` in the loop body as the code might not be auto-vectorized and CPU pipeline be broken.

Was there any reason you don't record the output in `flush()`?

Reply:
In the proposed fix, it says the record of output can be done in PagePartitioner.partitionPage.
Another reason that record of update is not in flush() is that in flush(), there is only outputBuffer of class OutputBuffer available, which seems not able to retrieve the page information.

Reply:
hi Sean, `recordOutput` takes two parameters, one for the positionCount (row count) for that partition, and one for the size in bytes for that partition. Each partition was represented using a PagePartitioner, and each PartitionBuffer has an array of BlockEncodingBuffers, one for each column. The positionCount is a member of PagePartitioner. For the size, you can sum up the serialized sizes of each BlockEncodingBuffer by calling getSerializedSizeInBytes().  This is because in OptimizedPartitionedOutputOperator, the pageBuilder/blockBuilder were replaced by the BlockEncodingBuffer so we need to get the sizes of the BlockEncodingBuffers instead of the Page built from the pageBuilder in the non-optimized version.

Reply:
Thanks Ying. But I don't see positionCount as a member of PagePartitioner in OptimizedPartitionedOutputOperator.

Reply:
About sum up the serialized sizes of each BlockEncodingBuffer by calling getSerializedSizeInBytes(). I tried it too.
        public void flush()
        {
            for (int i = 0; i < partitionBuffers.length; i++) {
                //partitionBuffers[i].flush(outputBuffer);
                long totalSizeInBytes = 0;
                for (int j = 0; j < partitionBuffers[i].blockEncodingBuffers.length; j++)
                    totalSizeInBytes += partitionBuffers[i].blockEncodingBuffers[j].getSerializedSizeInBytes();
                operatorContext.recordOutput(totalSizeInBytes, partitionBuffers[i].getPositionCount());
                partitionBuffers[i].flush(outputBuffer);
            }
        }

But it does not give accurate operatorContext.getOutputDataSize().getTotalCount(), no matter I put the recordOuput() before or after partitionBuffers[i].flush(outputBuffer), while putting it before and after flush() gives different results.

Reply:
Sean, for the number of rows, you can introduce a new field `outputPositionCount` in PagePartitioner to keep track of the number of output rows. For output data size, your test should reflect the actual output data size, not the other way around(i.e. write wrong or inefficient code to match the tests). The `getSizeInBytes` in `Block` is to estimate the logical block sizes. If the output data size is not the same as the input page/blocks size, then it is not the same. It is ok to use a range instead of a single value to verify the results. You can use `assertBetweenInclusive` or `assertBetweenExclusive` for that purpose. Can you post an example of the the results where you meant "inaccurate"? Was it able to be represented as a range e.g. [0.9, 1.1] * input page size?

Reply:
In the simple page test, where the page input is 900bytes, if I put recordOutput BEFORE partitionBuffers[i].flush(outputBuffer); the operatorContext.getOutputDataSize().getTotalCount() return 1104 bytes. If I put recordOutput AFTER partitionBuffers[i].flush(outputBuffer); the operatorContext.getOutputDataSize().getTotalCount() return 304 bytes

Reply:
This is expected and that's exactly what `flush` does. So we should report the numbers before the flush happens.

Reply:
Hi Ying,
If I add a new field `outputPositionCount` in PagePartitioner to keep track of the number of output rows, then I will need to update it in `partitionPage(Page page)` function, then the behavior will similar to the diff I have currently, correct?

For your question "Was it able to be represented as a range e.g. [0.9, 1.1] * input page size?". The difference of size when test with a simple page is around 10%. But if test with a page with DictionaryBlock or RLE block, the difference is very big (>50%).
```
        public void flush()
        {
            long totalSizeInBytes = 0;
            long totalPositionCount = 0;
            for (int i = 0; i < partitionBuffers.length; i++) {
                if (partitionBuffers[i].blockEncodingBuffers == null)
                    continue;
                for (int j = 0; j < partitionBuffers[i].blockEncodingBuffers.length; j++)
                    totalSizeInBytes += partitionBuffers[i].blockEncodingBuffers[j].getSerializedSizeInBytes();
                totalPositionCount += partitionBuffers[i].getPositionCount();
                partitionBuffers[i].flush(outputBuffer);
            }
            operatorContext.recordOutput(totalSizeInBytes, totalPositionCount); // totalPositionCount is wrong with multiple pages due to partitionBuffers[i].resetPositions() in partitionPage()
        }
```